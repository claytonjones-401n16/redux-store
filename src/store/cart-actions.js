import axios from 'axios';

export const removeFromCart = (payload) => {
  
  return async dispatch => {
    await axios.put(`https://js-401-lab-07.herokuapp.com/api/v1/products/${payload._id}`, {...payload, stock: payload.stock + 1});
    dispatch(actualRemove(payload));
  }
}
export const actualRemove = (payload) => {
  return {
    type: 'REMOVE_FROM_CART',
    payload
  }
}