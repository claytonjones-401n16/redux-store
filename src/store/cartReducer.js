export const initState = {
  cartCount: 0,
  cartContents: [],
}

export default (state = initState, action) => {
  let newState = { ...state };

  switch (action.type) {
    case 'ADD_TO_CART':
      newState.cartContents = [...newState.cartContents, action.payload];
      newState.cartCount++;
      break;
    case 'REMOVE_FROM_CART':
      for (let i = 0; i < newState.cartContents.length; i++) {
        if (action.payload.name === newState.cartContents[i].name) {
          newState.cartContents.splice(i, 1);
          newState.cartContents = [...newState.cartContents];

          newState.cartCount--;
          break;
        }
      }
      break;
    default:
      break;
  }

  return newState;
}