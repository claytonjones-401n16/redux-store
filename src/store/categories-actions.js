import axios from 'axios';

export const get = () => async dispatch => {
  let results = await axios.get('https://js-401-lab-07.herokuapp.com/api/v1/categories');
  dispatch(createInitialList(results.data.results));
}

export const createInitialList = (payload) => {
  return {
    type: 'GET_CATEGORIES',
    payload
  }
}

export const changeCategory = (payload) => {
  return {
    type: 'CHANGE_CATEGORY',
    payload
  }
}