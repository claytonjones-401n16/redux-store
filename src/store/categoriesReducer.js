export const initState = {
  allCategories: [],
  currentCategory: '',
};

export default (state = initState, action) => {
  let newState = { ...state };

  switch (action.type) {
      case 'GET_CATEGORIES':
        newState.allCategories = [...action.payload];
        break;
      case 'CHANGE_CATEGORY':
          newState.currentCategory = action.payload;
          break;
      case 'ADD_CATEGORY':
          newState.allCategories.push(action.payload);
          break;
      default:
          break;
  }

  return newState;
};

