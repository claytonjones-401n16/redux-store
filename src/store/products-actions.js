import axios from 'axios';

export const get = () => async dispatch => {
  let results = await axios.get('https://js-401-lab-07.herokuapp.com/api/v1/products')
  dispatch(createInitialList(results.data.results))
}

const createInitialList = (payload) => {
  return {
    type: 'GET_PRODUCTS',
    payload
  }
}

export const addToCart = (payload) => {
  return async dispatch => {
    dispatch(toggleAddButtons());
    await axios.put(`https://js-401-lab-07.herokuapp.com/api/v1/products/${payload._id}`, {...payload, stock: payload.stock - 1});
    dispatch(actualAdd(payload));
    dispatch(toggleAddButtons());
  }
}

export const actualAdd = (payload) => {
  return {
    type: 'ADD_TO_CART',
    payload
  }
}

const toggleAddButtons = () => {
  return {
    type: 'TOGGLE_ADD_BUTTONS'
  }
}
