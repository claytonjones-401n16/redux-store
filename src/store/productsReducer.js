export const initState = {
  allProducts: [],
  addButtonDisabled: false,
  productToView: {}
};

export default (state = initState, action) => {
  let newState = { ...state };

  switch (action.type) {
      // define our application's acceptable actions

      case 'GET_PRODUCTS':
        newState.allProducts = [...action.payload];
        break;
      case 'ADD_PRODUCT':
          newState.allProducts.push(action.payload);
          break;
      case 'ADD_TO_CART':
          for (let i = 0; i < newState.allProducts.length; i++) {
            if (action.payload.name === newState.allProducts[i].name) {
              newState.allProducts[i].stock = newState.allProducts[i].stock - 1;
              newState.allProducts = [...newState.allProducts];
              break;
            }
          }
          if (newState.productToView.name && newState.productToView.name === action.payload.name) {
            newState.productToView = {...newState.productToView, stock: newState.productToView.stock - 1}
          }
          break;
      case 'REMOVE_FROM_CART':
        for (let i = 0; i < newState.allProducts.length; i++) {
          if (action.payload.name === newState.allProducts[i].name) {
            newState.allProducts[i].stock = newState.allProducts[i].stock + 1;
            newState.allProducts = [...newState.allProducts];
            break;
          }
        }
        break;
      case 'TOGGLE_ADD_BUTTONS':
        newState = {...newState, addButtonDisabled: !newState.addButtonDisabled};
        break;
      case 'VIEW_ONE_PRODUCT':
        newState = {...newState, productToView: {...action.payload}}
        break;
      default:
          break;
  }

  return newState;
};

