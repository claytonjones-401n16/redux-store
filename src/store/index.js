// xx - define the initial state
// define the reducer
// -- defining the action types
// -- define the behavior for each action type
// redux magic to combine the init state and reducer into a redux store

// useState(initState)
// this.state = initState

import { createStore, combineReducers } from 'redux';
import cartReducer from './cartReducer';
import categoriesReducer from './categoriesReducer';
import productsReducer from './productsReducer';


const reducers = combineReducers({
    cart: cartReducer,
    categories: categoriesReducer,
    products: productsReducer
})

export default createStore(reducers);
