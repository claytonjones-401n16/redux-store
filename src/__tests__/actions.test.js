import axios from 'axios';
import nock from 'nock';
import thunk from '../store/middleware/thunk';
import configureMockStore from 'redux-mock-store';

import * as cartActions from '../store/cart-actions';
import * as categoriesActions from '../store/categories-actions';
import * as productsActions from '../store/products-actions';

jest.mock('axios');

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

describe('cart actions', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      cartCount: 0,
      cartContents: [],
    });
  });

  afterEach(() => {
    nock.cleanAll();
  });

  it('creates REMOVE_FROM_CART after API call', () => {
    nock('https://js-401-lab-07.herokuapp.com')
      .put('/api/v1/products/1234')
      .reply(200);

    return store.dispatch(cartActions.removeFromCart({_id: 1234}))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });
});

describe('categories actions', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      allCategories: [],
      currentCategory: '',
    });
  });

  afterEach(() => {
    nock.cleanAll();
  });

  it('creates GET_CATEGORIES after API call', () => {

    let categoriesData = {
      data: {
        results: ['cat1', 'cat2']
      }
    }

    // nock('https://js-401-lab-07.herokuapp.com')
    //   .get('/api/v1/categories')
    //   .reply(200, 'hello');

    // nock for some reason doesn't work here but the jest axios mock does
    axios.get.mockResolvedValue(categoriesData);

    return store.dispatch(categoriesActions.get())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      });
  });
});

describe('products actions', () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      allProducts: [],
    });
  });

  afterEach(() => {
    nock.cleanAll();
  });

  it('creates GET_PRODUCTS after API call', () => {
    let productsData = {
      data: {
        results: ['products', 'more products']
      }
    }

    axios.get.mockResolvedValue(productsData);

    return store.dispatch(productsActions.get())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      })
  })
});