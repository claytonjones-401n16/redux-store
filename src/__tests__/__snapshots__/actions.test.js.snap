// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart actions creates REMOVE_FROM_CART after API call 1`] = `
Array [
  Object {
    "payload": Object {
      "_id": 1234,
    },
    "type": "REMOVE_FROM_CART",
  },
]
`;

exports[`categories actions creates GET_CATEGORIES after API call 1`] = `
Array [
  Object {
    "payload": Array [
      "cat1",
      "cat2",
    ],
    "type": "GET_CATEGORIES",
  },
]
`;

exports[`products actions creates GET_PRODUCTS after API call 1`] = `
Array [
  Object {
    "payload": Array [
      "products",
      "more products",
    ],
    "type": "GET_PRODUCTS",
  },
]
`;
